sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_responses.hpp

```diff
@@ -9,102 +9,222 @@
 #include <vector>
 
 #include <azure/core/operation.hpp>
+#include <azure/storage/common/paged_response.hpp>
 
+#include "azure/storage/blobs/blob_options.hpp"
 #include "azure/storage/blobs/protocol/blob_rest_client.hpp"
 
-namespace Azure { namespace Storage { namespace Blobs {
+namespace Azure { namespace Storage {
+
+  namespace Files { namespace DataLake {
+    class ListFileSystemsPagedResponse;
+  }} // namespace Files::DataLake
```

Comment:
forward declaration. So that we can define friend class.

Reply:
> forward declaration.

This is certainly odd though, do we have namespace mix-and-match outside of each package already? I understand a lot of the **implementation details** of datalake is based on blobs, but we shouldn't muddy the package boundaries like this from a public surface area. The `DataLake` namespace and types in that namespace should only be found in the datalake package. @JeffreyRichter and @RickWinter do you agree or am I off-base here.

> So that we can define friend class.

I thought we could get rid of those.

If there is no way for you to avoid this/work around it for now, I suppose we could defer addressing it after beta.10, and discuss further.

Reply:
Using forward declaration doesn't mean blob package depends on datalake package.

Reply:
We can keep discussing, but it's not a breaking change. it's all about internal implementation.

Reply:
> Using forward declaration doesn't mean blob package depends on datalake package.

Hmm... the `ListFileSystemsPagedResponse` is defined in the datalake package, correct, within `datalake_responses.hpp`? Why doesn't that invert the dependency?

If I just install the blobs package, without datalake, and call` ListBlobContainersPagedResponse ListBlobContainers();`or reference directly `ListFileSystemsPagedResponse`, would my application code build and run correctly?

Reply:
> Hmm... the ListFileSystemsPagedResponse is defined in the datalake package, correct, within datalake_responses.hpp? Why doesn't that invert the dependency?

Forward declaration doesn't require the type to be actually defined.

```C++
class MyType;  // this is a forward declaration

MyType* ptr;  // This is fine, because it's just a pointer, compiler allocates 4 or 8 bytes here, and that's it. Compiler doesn't need to know anything else.

friend class MyType;  // This is also fine

MyType.func();  // This is not fine. Compiler has to know the definition of MyType.
MyType.member  // likewise
```

Reply:
> If I just install the blobs package, without datalake, and call ListBlobContainersPagedResponse ListBlobContainers();

It's fine.

> or reference directly ListFileSystemsPagedResponse

This won't work. But we only declare `ListFileSystemsPagedResponse` a friend class. So there's no way we can reference it.